38-allowlist-management.md

```diff
@@ -0,0 +1,162 @@
+- [ ] Test plan is in place
+- [ ] User-facing documentation has been created in [keylime/keylime-docs]
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+## Summary
+
+Allowlist management is, obviously, managing allowlists. Currently the allowlist
+is binded with agent records, there are two issues in current design:
+
+* Allowlist reusing: An allowlist can't be shared between agents, so we can't apply
+  the same data to identical environments. By support reusing, we also can provide
+  batch allowlist updates for matching agents.
+* Performance: Since the ima measurements could be very large, it has impact on query
+  efficiency when the allowlist does not participate in the agent processing, e.g., we
+  don't need to load the allowlist for updating agent state.
+
+
+## Motivation
+
+### Goals
+
+The goal is to provide allowlist management API and CLI for allowlist management,
+establish relationship between an allowlist and an agent. The allowlist used by agent
+can be changed when the host of agent is upgraded and the old allowlist doesn't apply.
+
+### Non-Goals
+
+Import allowlist from agent.
+
+## Proposal
+
+The proposal is to:
+
+* Create a new table named ``allowlists`` to hold allowlist data that original hosted
+  in the ``verifiermain`` table, namely: ``tpm_policy``, ``vtpm_policy``, ``and allowlist``.
+* Add a foreign key named ``allowlist_id`` in the ``verifiermain`` table to reference the
+  allowlist data that will be used for the agent.
+* Implements ``/v2/allowlists/`` API for Verifier to support CRD operations. Update an
+  allowlist is not included in the initial proposal but can be implemented if necessary.
+* Enhance ``tenant`` CLI to manage allowlists, when adding agent to the Verifier, user
+  specifies an existing ``allowlist`` name instead of allowlist data to complete the
+  registration.
+
+
+### User Stories (optional)
+
+
+#### Story 1
+
+A user can reuse allowlist by:
+
+* Create an allowlist record by posting policies to the Verifier.
+* Add an agent to the Verifier by providing Agent ID, IP Address and the allowlist name
+  the agent will use.
+* Add another agent running under the same environment to the Verifier with the same allowlist.
+
+#### Story 2
+
+A user can update agent with new allowlist by:
+
+* Create an allowlist record by posting new policies to the Verifier.
+* Update an agent to the Verifier by updating the allowlist name.
+* Agent will use new allowlist to verify system integrety.
+
+### Notes/Constraints/Caveats (optional)
+
+None
+
+### Risks and Mitigations
+
+None
+
+## Design Details
+
+DB Changes:
+
+The ``allowlists`` table will be created as:
+
+```
+op.create_table('allowlists',
+                sa.Column('id', sa.Integer(), nullable=False),
+                sa.Column('name', sa.String(255), nullable=False),
+                sa.Column('tpm_policy', sa.Text(), nullable=True),
```

Comment:
Should we specify the lengths of the tpm_policy and vtpm_policy fields? Otherwise will it default to whatever the specific engine defaults to so different installs would be slightly different? Just wanting to avoid some weird possible issues in the future.

Reply:
It depends on the backend, for mysql the TEXT defaults to 16k or so, and postgresql has no strict size limit. I am not sure about sqlite, but it seems supporting up to 2G.
We usually don't specifiy a specific size for text fields, unless the data we are holding exceeds the limit at some day, but it's a rare case, since by design tpm_policy and vtpm_policy should only occupy a small amount of data.
Not specifiying a hard lengh at db layer is to avoid frequent db change, we can't anticipate everything. If we want to apply a size limit, it can be done at API level, I would avoid to pass through and let db complain :)

Reply:
A scenario I've encountered before with a system that supports multiple databases: The length isn't specified and the data in a field keeps getting larger and larger. The same data works on some setups but not others because the lengths weren't specified. Tracking down why something works on one system but not another (same versions of the software, same data) are much harder than an error about a data field being too large. I'm in favor of being specific for data field lengths, even if it could mean more db migrations in the future because at least we're being specific.

Reply:
I know the length is valid for varchar, but not sure it also works for TEXT, it appears to me that the length we specified is not a hard size limit under the hood, more like a hint to the sqlalchemy that which type of TEXT will be used.
There are three types of TEXT in mysql:
TEXT ~64kb
MEDIUMTEXT ~16Mb
LONGTEXT ~4Gb
I didn't use sqlite much in the past, but from the shema it's a longtext field for current allowlist, so I suppose it's similar to mysql.